src/openms/source/ANALYSIS/ID/AhoCorasickAmbiguous.cpp

```diff
@@ -80,7 +80,7 @@ namespace OpenMS
   void ACTrie::addNeedle(const std::string& needle)
   {
     Index cn {0}; // start at root
-    for (auto c : needle)
+    for (const auto& c : needle)
```

Comment:
this is almost certainly a pessimisation, since its only a single byte.
We had a comment keyword to annotate that a simple `auto` is ok, right?

Reply:
I checked this for a few cases in godbolt, and it didn't use a pointer/similar but was always able to generate the same asm code.

Reply:
Alternatively, just use char c here instead of auto then it is also clear that a copy doesn't hurt

Reply:
> I checked this for a few cases in godbolt, and it didn't use a pointer/similar but was always able to generate the same asm code.

The asm generated highly depends on the context: the compiler must re-read an `a&` value from memory everytime it is mentioned, since it cannot know if the value changed in the meantime (keyword: aliasing). So copy is the safe bet and allows for more optimizations (potentially).

Reply:
> Alternatively, just use char c here instead of auto then it is also clear that a copy doesn't hurt

Not sure here, since this might incur a conversion, if some implementations use `unsigned char` vs `char`. Auto is the safe bet.

Reply:
A std::string is a std::basic_string<char>

Reply:
yep, correct. The time it takes to figure out that this is guaranteed by the standard is probably not worth arguing over 'auto' vs. 'char'?  ;)

Reply:
agreed, ;) why I am nitpicking: it would have the nice side effect that we could just have the guideline to use mainly const auto& in the general case or the actual type for the build in ones. But not so important, I think this is done...